go/dbagent/dbagent.go

```diff
@@ -0,0 +1,360 @@
+}
+
+func GetMySQLInfo() (Info MySQLInfo, err error) {
+	Info.MySQLVersion, err = GetMySQLVersion()
+	Info.MySQLDatadirPath, err = getMySQLDatadirPath()
+	Info.MySQLBackupdirPath = config.Config.MySQLBackupDir
+	Info.IsSlave, err = isSlave()
+	Info.IsMaster, err = isMaster()
+	Info.IsBinlogEnabled, err = isBinlogEnabled()
+	Info.HasActiveConnections, err = hasActiveConnections()
+	return Info, err
+}
+
+func GetMySQLDatabases() (Databases []string, err error) {
+	query := `SELECT SCHEMA_NAME FROM information_schema.schemata WHERE SCHEMA_NAME NOT IN ('information_schema','mysql','performance_schema','sys');`
+	err = QueryData(query, sqlutils.Args(), func(m sqlutils.RowMap) error {
+		db := m.GetString("SCHEMA_NAME")
+		Databases = append(Databases, db)
+		return nil
+	})
+	if err != nil {
+		log.Errore(err)
+	}
+	return Databases, err
+}
+
+func GetMySQLEngines(dbname string) (Engines []string, err error) {
+	query := `SELECT engine FROM information_schema.tables where TABLE_SCHEMA = ? and table_type = 'BASE TABLE' GROUP BY engine;`
+	err = QueryData(query, sqlutils.Args(dbname), func(m sqlutils.RowMap) error {
+		engine := m.GetString("engine")
+		Engines = append(Engines, engine)
+		return nil
+	})
+	if err != nil {
+		log.Errore(err)
+	}
+	return Engines, err
+}
+
+func GetTokuDBSize(dbname string) (TokuSize int64, err error) {
+	query := `SELECT SUM(bt_size_allocated) AS tables_size FROM information_schema.TokuDB_fractal_tree_info WHERE table_schema = ?;`
+	err = QueryData(query, sqlutils.Args(dbname), func(m sqlutils.RowMap) error {
+		TokuSize = m.GetInt64("tables_size")
+		return nil
+	})
+	if err != nil {
+		log.Errore(err)
+	}
+	return TokuSize, err
+}
+
+func GetInnoDBLogSize() (InnoDBLogSize int64, err error) {
+	query := `SELECT @@innodb_log_file_size*@@innodb_log_files_in_group AS logFileSize;`
+	err = QueryData(query, sqlutils.Args(), func(m sqlutils.RowMap) error {
+		InnoDBLogSize = m.GetInt64("logFileSize")
+		return nil
+	})
+	if err != nil {
+		log.Errore(err)
+	}
+	return InnoDBLogSize, err
+}
+
+func UserExists(username string) (UserExists bool) {
+	query := `SELECT user FROM mysql.user WHERE user=?;`
+	err := QueryData(query, sqlutils.Args(username), func(m sqlutils.RowMap) error {
+		UserExists = (len(m.GetString("user")) != 0)
+		return nil
+	})
+	if err != nil {
+		log.Errore(err)
+	}
+	return UserExists
+}
+
+func HasGrant(username string, grant string) (HasGrant bool) {
+	query := `SELECT * FROM mysql.user WHERE user=?;`
+	err := QueryData(query, sqlutils.Args(username), func(m sqlutils.RowMap) error {
+		HasGrant = (m.GetString(grant) == "Y")
+		return nil
+	})
+	if err != nil {
+		log.Errore(err)
+	}
+	return HasGrant
+}
+
+func CreateUser(username string, host string, password string) error {
+	_, err := ExecuteQuery(`CREATE USER IF NOT EXISTS ?@? IDENTIFIED BY ?;`, username, host, password)
+	return err
+}
+
+func GrantUser(username string, host string, grant string, object string, password string) error {
```

Comment:
Why does `orchestrator-agent` need to issue a `GRANT` command?

Reply:
I advise against this.

Reply:
I agree with your point that user provisioning and management shouldn't be done by orchestrator agent. Initially, this code was used for provision replication user and permissions for it. But there are chef\puppet or other configuration management tools for these cases. So i've removed it in b73ea71